### ISSUE TITLE
Combined data frame with data/predicted/resid

### ISSUE THREAD

**Issue by bwiernik** [easystats/insight#342]
One thing that `broom::augment()` makes easy that seems difficult currently with insight is to get a combined data frame with the input data, predictions (+SEs/intervals), and residuals. Such a data frame is potentially useful for:

1. Constructing plots that overlay raw data with model predictions.
2. Detailed model diagnositics, such as examinate of residual correlations with individual predictors (versus only against the fitted values as in `check_model()` plots).

Do you think we could add a function that calls get_data(), get_predicted(), and get_data() and merging them by rowname? Thoughts on a name? 

It would be possible of course for users to that manually, but I'm thinking about cases like my intro regression students where I try to limit the amount of unique R data processing steps I introduce at once. This is especially an issue with things like missing data that might cause rows not to be aligned across various outputs, necessitating merging rather than cbinding (e.g., https://github.com/easystats/insight/issues/341).

The other things that augment() includes are diagnostic statistics hat, cooksd, standardized residuals, and leave-one-out estimates (sigma, dffits, dfbeta). I don't see much need for these combined with predicted/resid/raw data, but it might be useful to export a function for that data frame in performance outside of the internal functions used by `check_model()`.

**Comment by strengejacke** [comment]
You mean something like this?

``` r
library(insight)
m <- lm(mpg ~ hp + cyl, data = mtcars)
cbind(get_data(m), get_predicted(m), Residuals = get_residuals(m))
#>                      mpg  hp cyl Predicted        SE    CI_low  CI_high
#> Mazda RX4           21.0 110   6  21.21678 0.7281647 19.727518 22.70605
#> Mazda RX4 Wag       21.0 110   6  21.21678 0.7281647 19.727518 22.70605
#> Datsun 710          22.8  93   4  26.07124 0.9279509 24.173366 27.96911
#> Hornet 4 Drive      21.4 110   6  21.21678 0.7281647 19.727518 22.70605
#> Hornet Sportabout   18.7 175   8  15.44448 0.9200310 13.562810 17.32616
#> Valiant             18.1 105   6  21.31239 0.7777664 19.721680 22.90310
#> Duster 360          14.3 245   8  14.10597 1.0080670 12.044237 16.16769
#> Merc 240D           24.4  62   4  26.66401 0.9225132 24.777260 28.55076
#> Merc 230            22.8  95   4  26.03299 0.9362657 24.118117 27.94787
#> Merc 280            19.2 123   6  20.96820 0.6234320 19.693139 22.24326
#> Merc 280C           17.8 123   6  20.96820 0.6234320 19.693139 22.24326
#> Merc 450SE          16.4 180   8  15.34888 0.8862558 13.536280 17.16147
#> Merc 450SL          17.3 180   8  15.34888 0.8862558 13.536280 17.16147
#> Merc 450SLC         15.2 180   8  15.34888 0.8862558 13.536280 17.16147
#> Cadillac Fleetwood  10.4 205   8  14.87083 0.8057154 13.222961 16.51871
#> Lincoln Continental 10.4 215   8  14.67962 0.8206255 13.001249 16.35798
#> Chrysler Imperial   14.7 230   8  14.39279 0.8911693 12.570146 16.21544
#> Fiat 128            32.4  66   4  26.58752 0.9099596 24.726448 28.44860
#> Honda Civic         30.4  52   4  26.85523 0.9695585 24.872258 28.83820
#> Toyota Corolla      33.9  65   4  26.60665 0.9127445 24.739874 28.47342
#> Toyota Corona       21.5  97   4  25.99475 0.9454598 24.061069 27.92843
#> Dodge Challenger    15.5 150   8  15.92253 1.1490264 13.572504 18.27255
#> AMC Javelin         15.2 150   8  15.92253 1.1490264 13.572504 18.27255
#> Camaro Z28          13.3 245   8  14.10597 1.0080670 12.044237 16.16769
#> Pontiac Firebird    19.2 175   8  15.44448 0.9200310 13.562810 17.32616
#> Fiat X1-9           27.3  66   4  26.58752 0.9099596 24.726448 28.44860
#> Porsche 914-2       26.0  91   4  26.10948 0.9205392 24.226768 27.99220
#> Lotus Europa        30.4 113   4  25.68880 1.0474287 23.546572 27.83104
#> Ford Pantera L      15.8 264   8  13.74265 1.2011595 11.286007 16.19930
#> Ferrari Dino        19.7 175   6  19.97387 0.7635547 18.412227 21.53552
#> Maserati Bora       15.0 335   8  12.38501 2.1153615  8.058613 16.71141
#> Volvo 142E          21.4 109   4  25.76529 1.0175965 23.684073 27.84651
#>                      Residuals
#> Mazda RX4           -0.2167823
#> Mazda RX4 Wag       -0.2167823
#> Datsun 710          -3.2712383
#> Hornet 4 Drive       0.1832177
#> Hornet Sportabout    3.2555152
#> Valiant             -3.2123908
#> Duster 360           0.1940339
#> Merc 240D           -2.2640109
#> Merc 230            -3.2329949
#> Merc 280            -1.7682002
#> Merc 280C           -3.1682002
#> Merc 450SE           1.0511237
#> Merc 450SL           1.9511237
#> Merc 450SLC         -0.1488763
#> Cadillac Fleetwood  -4.4708339
#> Lincoln Continental -4.2796170
#> Chrysler Imperial    0.3072085
#> Fiat 128             5.8124759
#> Honda Civic          3.5447721
#> Toyota Corolla       7.2933542
#> Toyota Corona       -4.4947515
#> Dodge Challenger    -0.4225272
#> AMC Javelin         -0.7225272
#> Camaro Z28          -0.8059661
#> Pontiac Firebird     3.7555152
#> Fiat X1-9            0.7124759
#> Porsche 914-2       -0.1094817
#> Lotus Europa         4.7111956
#> Ford Pantera L       2.0573462
#> Ferrari Dino        -0.2738720
#> Maserati Bora        2.6149866
#> Volvo 142E          -4.3652912
```

<sup>Created on 2021-04-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by bwiernik** [comment]
Yep

**Comment by DominiqueMakowski** [comment]
Adding data with the predictions in a dataframe was something that `modelbased` does (the initial goal of `get_predicted()` was to be a robust wrapper around `predict()` which returns a vector of predictions, rather than a df). Maybe the residuals could be added by default for `modelbased::estimate_predictions()` (that generate predictions on the original dataframe by default), which implicitly are rather used to do [things like this](https://easystats.github.io/modelbased/articles/estimate_response.html#prediction-against-original-data)

**Comment by bwiernik** [comment]
I'm not following what you are saying. `predict()` doesn't produce a vector if se.fit or interval are specified.

**Comment by DominiqueMakowski** [comment]
Yes, I was referring to its default output. The output is indeed quite inconsistent; when se.fit is set to true it returns a list, and a df if only interval is s(at least for lm). Currently we went for consistency with `get_predicted` by always sticking to a vector as the default primary output (which is quite convenient) that can be transformed as a dataframe explicitly (using `as.data.frame()`)

**Comment by bwiernik** [comment]
Okay, what's your broader thinking about including a function like this in insight versus modelbased?

**Comment by DominiqueMakowski** [comment]
Dealing with all the predictions, the intervals and all is quite complex, + it's closely related to marginal estimations (emmeans-style), as well as visualization. These 3 aspects were the reason why originally we thought it would be good to have a separate package. This package would have been dedicated to generating predictions and computing interval, and importantly could have some specific dependencies (e.g., emmeans). It would also be closely connected to visualization and things like data-grids creation. I still think that these aspects warrant having a dedicated package, and given the shift towards ["generative modeling"](https://twitter.com/Nate__Haines/status/1297912972707352576), a package dedicated to model-based estimations makes sense.

However, at some point when implementing some tests in `performance`, we realized that it'd be nice to have a little insight `predict()` wrapper (as `predict` doesn't work for all models). One thing lead to another this started to grow with more exceptions and all, and at some point we realised that it's weird to have this predict wrapper that is unable to provide the intervals (which could be also useful in other packages - since we see `insight` as a cornerstone brick of the ecosystem), so we decided to port all the pure predictive machinery to insight for convenience. At its core, it's really meant to be a robust (and improved) alternative/wrapper over `predict()`, that can be used for different high-level purposes, such as modelbased ones.

That said, comparatively to the other packages, modelbased is now a bit behind in terms of development, functionalities and love. At some point it was doing great, helped by its narrower scope (imo it had some of the best vignettes of the easyverse), but then it stagnated (it's mostly my fault I feel like time's just slipping through my fingers at an incredible speed). But 'easystats' seems to have a cyclical development movement (actually @strengejacke we should plot the github activity of all the packages over time it would be quite interesting), so I have no doubts we will come back to `modelbased` once other priorities are dealt with (e.g., "report")

**Comment by strengejacke** [comment]
what could be a good function name for this?

**Comment by DominiqueMakowski** [comment]
for this what?

**Comment by strengejacke** [comment]
> Combined data frame with data/predicted/resid

**Comment by DominiqueMakowski** [comment]
`modelbased::estimate_predictions`?

**Comment by DominiqueMakowski** [comment]
I.e., I think that it's better not to disperse ourselves API-wise, `insight` should focus on providing the basic primitives (residuals, predictions, data...) and other higher-level packages can do integrations of these elements, adapted to a given use-case.
